added blog SC

windushka opened the pull request.

qdrn commented on .github/workflows/ci-tests_blog.yml:
```diff
@@ -0,0 +1,26 @@
+name: Unit tests
```
I believe the naming of the file is unfortunate. Either you should rename the file that test the helloworld as well. Either you should put everything under `ci-tests.yml`.

qdrn commented on blog/assembly/__tests__/massa-example.spec.ts:
```diff
@@ -0,0 +1,19 @@
+  it('should not throw with no arguments', () => {
+    // We're using the expect function to make an assertion about the behavior of the `constructor` function.
+    // Specifically, we're expecting that calling `main` with serialized `constructor` will not throw an error.
+    expect(
```
You should test:
- constructor
- post
- deletePost
- _blogKey

qdrn commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,44 @@
+    Storage.set("N_BLOG_POSTS", n_blog_posts.toString());
+}
+
+export function deletePost(binaryArgs: StaticArray<u8>): void {
```
Should improve documentation

qdrn commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,44 @@
+    Storage.set(_blogKey(postIndex), "");
+}
+
+function _blogKey(postIndex: string): string {
```
Same

qdrn commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,44 @@
+    return [];
+}
+
+export function post(binaryArgs: StaticArray<u8>): void {
```
Should improve documentation

qdrn commented:
Also you should fix the linter!

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,53 @@
+export function constructor(_: StaticArray<u8>): StaticArray<u8> {
+  // This line is important. It ensures that this function can't be called in the future.
+  // If you remove this check, someone could call your constructor function and reset your smart contract.
+  if (!callerHasWriteAccess()) {
```
You should be able to use [isDeployingContract](https://github.com/massalabs/massa-as-sdk/blob/main/assembly/std/context.ts#L46) which is cleaner

gregLibert replied:
Why was this discussioin resolved ? isDeployingContract is not working ?

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,53 @@
+  return [];
+}
+
+export function post(binaryArgs: StaticArray<u8>): void {
+  const argsDeser = new Args(binaryArgs);
```
pratice is to use _args and args:

```suggestion
export function post(_args: StaticArray<u8>): void {
  const args = new Args(_args);
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,53 @@
+export function post(binaryArgs: StaticArray<u8>): void {
+  const argsDeser = new Args(binaryArgs);
+  // Get the post from the arguments and check if it is valid (not null)
+  const post = argsDeser
```
```suggestion
  const postContent = args
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,53 @@
+  const post = argsDeser
+    .nextString()
+    .expect('Post argument is missing or invalid');
+  let nBlogPosts = parseInt(Storage.get("N_BLOG_POSTS"));
```
```suggestion
  let numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS"));
```

gregLibert commented on blog/package.json:
```diff
@@ -0,0 +1,45 @@
+{
+    "name": "my-massa-sc",
```
it should be something around blog and examples no ?

gregLibert reviewed (changes requested):
Good job, small improvments to get the code easier to understand and the tool easier to maintain.

windushka commented on .github/workflows/ci-tests_blog.yml:
```diff
@@ -0,0 +1,26 @@
+name: Unit tests
```
Done

windushka commented on .github/workflows/daily-deploy_blog.yml:
```diff
@@ -0,0 +1,41 @@
+name: Daily Deploy
```
Done

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,53 @@
+  let numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS"));
+  numberOfPosts += 1;
+  // Store the post in the storage of the contract with the key POST_postIndex
+  // The keys will have the following syntaxes: POST_1.0, POST_2.0, POST_3.0, etc.
```
I'm not sure about the .0 in 1.0, 2.0 ... numberOfPosts is an integer so it should be POST_1, POST_2 ...

windushka replied:
It is indeed an integer but for some unusual reason numberOfPosts will always have the decimal part even when applying Math.floor() on it. 

![image](https://user-images.githubusercontent.com/92168447/231753270-4f8a61c9-f166-42a7-b605-533dc3014610.png)

windushka replied:
For isDeployingContract() to work I had to update to the latest massa-as-sdk version which resulted in my unit tests not working. When readingthe storage the returned value would always be NaN. I've addressed this issue to the run team.

gregLibert replied:
This is not a priority. You can safely ignore my comment

windushka replied:
Using [isDeployingContract](https://github.com/massalabs/massa-as-sdk/blob/main/assembly/std/context.ts#L46) now and fixed the decimal part with the help of @Ben-Rey

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @param _ - not used
+ */
+export function constructor(_: StaticArray<u8>): StaticArray<u8> {
```
```suggestion
export function constructor(_: StaticArray<u8>): void {
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+  // This line is important. It ensures that this function can't be called in the future.
+  // If you remove this check, someone could call your constructor function and reset your smart contract.
+  if (!Context.isDeployingContract()) {
+    return [];
```
```suggestion
    return;
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+  Storage.set("N_BLOG_POSTS", nBlogPosts.toString());
+  // This function is used to emit an event to the blockchain
+  generateEvent('Blog initiated');
+  return [];
```
```suggestion
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+    .nextString()
+    .expect('Post argument is missing or invalid');
+
+  let numberOfPosts = 0;
```
```suggestion
  let postLastIndex = 0;
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+
+  let numberOfPosts = 0;
+  if (Storage.get<string>("N_BLOG_POSTS") !== "") {
+    numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS")) as i32;
```
```suggestion
    postLastIndex = parseInt(Storage.get("N_BLOG_POSTS")) as i32;
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+  if (Storage.get<string>("N_BLOG_POSTS") !== "") {
+    numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS")) as i32;
+  }
+  numberOfPosts += 1;
```
```suggestion
  postLastIndex += 1;
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+  numberOfPosts += 1;
+  // Store the post in the storage of the contract with the key POST_postIndex
+  // The keys will have the following syntaxes: POST_1, POST_2, POST_3, etc.
+  Storage.set(blogKey(numberOfPosts.toString()), post);
```
```suggestion
  Storage.set(blogKey(postLastIndex.toString()), post);
```

gregLibert commented on blog/assembly/contracts/main.ts:
```diff
@@ -0,0 +1,55 @@
+  // The keys will have the following syntaxes: POST_1, POST_2, POST_3, etc.
+  Storage.set(blogKey(numberOfPosts.toString()), post);
+  // Incrementing the value of N_BLOG_POSTS in the storage of the contract
+  Storage.set("N_BLOG_POSTS", numberOfPosts.toString());
```
```suggestion
  Storage.set("N_BLOG_POSTS", postLastIndex.toString());
```

windushka commented:
Addressed all the requested changes and added one max parallel job in the matrix strategy. When there were multiple deployment jobs running in parallel, sometimes the event wouldn't be received. Should be fixed now
